yaml_edit/lib/src/editor.dart

```diff
@@ -197,7 +197,7 @@ class YamlEditor {
   /// print(newNode.value); // "YAML"
   /// print(node.value); // "YAML Ain't Markup Language"
   /// ```
-  YamlNode parseAt(Iterable<Object> path, {YamlNode Function() orElse}) {
+  YamlNode? parseAt(Iterable<Object> path, {YamlNode? Function()? orElse}) {
```

Comment:
This change wasn't necessary to make the tests conform to null safety. However, I made the changes to conform to the documented interface saying:
```dart
  /// To get `null` when [path] does not point to a value in the [YamlNode]-tree,
  /// simply pass `orElse: () => null`.
```
This would mean that it's possible to get `null` as a result. When I did this change, some private methods needed to be changed to be null safe. I didn't change the private methods just so I can focus on the tests but this change made it necessary to change some of the tests to conform with the nullable result.

Reply:
Yeah, we're clearly using `null` as a sentinel value here.. Sometimes it's attractive to make a sentinel value, like a special instance of `YamlNode` that represents "no value". But I don't this is a good place..

The following is a value YAML document:
```yaml
key: value
otherkey: null
```
So we probably have to able to return null, because that could also be the value at the given path.

Reply:
What happens today, if you call `yamlEditor.parseAt(['key1', 'key2'])` on a document like:
```yaml
key1:
  key2: null
```

Does it return a `YamlNode` with `YamlNode.value == null`, I think not, but please do test it.

Reply:
@asartalo We need to find out of every place that takes or returns a `YamlNode` can return/take `null`.
I think the answer to that follows from `package:yaml` behavior, so I suggest exploring what that is.

Reply:
> What happens today, if you call yamlEditor.parseAt(['key1', 'key2']) on a document like:
>
> ```yaml
> key1:
>   key2: null
> ```
>
> Does it return a `YamlNode` with `YamlNode.value == null`, I think not, but please do test it.

@jonasfj Yes, right now on stable it does return a `YamlNode` with `YamlNode.value == null`. To be sure, I added some tests to demonstrate this on a clean master branch and found they pass. You can see this on PR #110 just so they are covered.

`YamlEditor.parseAt()` seems to guarantee to return `YamlNode` unless for exceptions (path errors, alias errors), or `null` when passed with `orElse: () => null` as described in the [documentation](https://pub.dev/documentation/yaml_edit/latest/yaml_edit/YamlEditor/parseAt.html). This is different from when we're accessing values directly using `package:yaml` like in the following:

```dart
import 'package:yaml/yaml.dart'; // yaml: ^3.1.0 (null safety)

void main(List<String> arguments) {
  const yamlSource = r'''
a:
  i: null
b:
  i: foo
''';

  final doc = loadYamlNode(yamlSource) as YamlMap;

  final value = doc['a']['i'];
  print('${value.runtimeType} - $value'); // Null - null

  final value2 = doc['b']['i'];
  print('${value2.runtimeType} - $value2'); // String - foo
}
``` 
In comparison, `YamlEditor.parseAt()` at current version (1.0.3), "always" returns a `YamlNode`.

```dart
import 'package:yaml_edit/yaml_edit.dart'; // yaml_edit: ^1.0.3

void main(List<String> arguments) {
  const yamlSource = r'''
a:
  i: null
b:
  i: foo
''';

  final doc = YamlEditor(yamlSource);

  final node = doc.parseAt(['a', 'i']);
  print('${node.runtimeType} - ${node.value}'); // YamlScalar - null

  final node2 = doc.parseAt(['b', 'i']);
  print('${node2.runtimeType} - ${node2.value}'); // YamlScalar - foo
}
```

`YamlEditor` achieves this by accessing the nodes directly.

The relevant lines are the following ([line 490-491 on master > editor.dart](https://github.com/google/dart-neats/blob/5ee7672e76b8303c6a33d002b5c1255dcc93ebf0/yaml_edit/lib/src/editor.dart#L490)):

```dart
      } else if (currentNode is YamlMap) {
        final map = currentNode as YamlMap;

        // ...snip...
        final keyNode = getKeyNode(map, keyOrIndex);

        // ...snip...
        currentNode = map.nodes[keyNode];
      } else {
```

In the above, `map.nodes` is of type `Map<dynamic, YamlNode>` and `currentNode` is always returned when found (except for `orElse` that returns null).

I've added some tests to cover this behavior for `YamlEditor.parseAt()` on my PR. If this is unclear, let me know.

Reply:
Perfect, this is a good find! :smile: 

Let's make the signature: `YamlNode parseAt(Iterable<Object?> path, YamlNode Function() orElse)`
Then we still have `orElse`, but it has to return a `YamlNode`.

If we want, we can always add a `parseAtOrNull` function, but that's hardly very useful, since a user will likely have to check if the response is an `YamlScalar` with `.value == null` anyways.

Reply:
@asartalo  actually, can't we make the signature:
`YamlNode parseAt(Iterable<Object?> path, {YamlNode? Function() orElse?})`

Then, if `orElse` returns `null`, we just wrap it in a `YamlScalar`, and return an `YamlScalar` instance with `.value == null`?

That way, a user calling `parseAt(['key'], orElse: () => null)` will get the same result whether the document is:
```yaml
{key: null}   # key is present, but the value is "null"
```
OR:
```yaml
{}  # key is not present
```
Which is likely what the user actually wants.

Reply:
Or we could make the signature:
`YamlNode parseAt(Iterable<Object?> path, {YamlNode Function() orElse?})`

And then simply update the documentation to not suggest:
```
final value = myYamlEditor.valueAt(['invalid', 'path'], orElse: () => null);
```
But instead suggest:
```
final value = myYamlEditor.valueAt(['invalid', 'path'], orElse: () => wrapAsYamlNode(null));
```

In fact, I think it's probably a mistake that we don't recommend that today, because the idea with `orElse: () => null` was that you wanted the same result regardless of whether the key was `null` or not present. This pattern of treating `null` and omitted key as the same is a very common and reasonable pattern :D

So I think you could argue that allowing `orElse` to return `null` invites bugs, because it makes people think they are handling the case where the value is at `path` in the YAML document is `null`, which is not what they are handling.

Always lovely to catch a bug when migrating to null-safety :rocket:

Reply:
I agree. It also one less codepath to remember as a user of the library and as a maintainer. I like the `YamlNode parseAt(Iterable<Object?> path, {YamlNode Function() orElse?})` signature . I'll add those changes to the documentation as well.